runtime/ints.c

```diff
@@ -849,3 +850,195 @@ CAMLprim value caml_nativeint_of_string(value s)
+                                     10000000000,
+                                     100000000000,
+                                     1000000000000,
+                                     10000000000000,
+                                     100000000000000,
+                                     1000000000000000,
+                                     10000000000000000,
+                                     100000000000000000,
+                                     1000000000000000000};
+  // We do not need to go further as OCaml only consider signed int64
+  //                                 10000000000000000000};
+  int log2 = u64_fast_log2(x);
+  int log10 = (1233 * log2) >> 12; // |log10(2) - 1233/4096| ~= 0.0000046
+  log10 += ((log10 < 18) && (x >= table[log10]));
+  return log10 + 1;
+}
+
+void write_digits(uint64_t val, char* buffer, size_t len) {
+  char* index = buffer + (len - 1); // writing from the end
+  // "00", "01", "02", ...
+  static const char DIGITS[200] = {
+    '0','0', '0','1', '0','2', '0','3', '0','4', '0','5', '0','6', '0','7',
+    '0','8', '0','9', '1','0', '1','1', '1','2', '1','3', '1','4', '1','5',
+    '1','6', '1','7', '1','8', '1','9', '2','0', '2','1', '2','2', '2','3',
+    '2','4', '2','5', '2','6', '2','7', '2','8', '2','9', '3','0', '3','1',
+    '3','2', '3','3', '3','4', '3','5', '3','6', '3','7', '3','8', '3','9',
+    '4','0', '4','1', '4','2', '4','3', '4','4', '4','5', '4','6', '4','7',
+    '4','8', '4','9', '5','0', '5','1', '5','2', '5','3', '5','4', '5','5',
+    '5','6', '5','7', '5','8', '5','9', '6','0', '6','1', '6','2', '6','3',
+    '6','4', '6','5', '6','6', '6','7', '6','8', '6','9', '7','0', '7','1',
+    '7','2', '7','3', '7','4', '7','5', '7','6', '7','7', '7','8', '7','9',
+    '8','0', '8','1', '8','2', '8','3', '8','4', '8','5', '8','6', '8','7',
+    '8','8', '8','9', '9','0', '9','1', '9','2', '9','3', '9','4', '9','5',
+    '9','6', '9','7', '9','8', '9','9'};
+
+  // Decode 4 digits at a time
+  while (val >= 10000) {
+    uint64_t r = val % 10000;
+    val /= 10000;
+    uint64_t r1 = 2 * (r / 100);
+    uint64_t r2 = 2 * (r % 100);
+
+    *index-- = DIGITS[r2 + 1];
+    *index-- = DIGITS[r2];
+    *index-- = DIGITS[r1 + 1];
+    *index-- = DIGITS[r1];
+  }
+
+  // Decode 2 digits at a time
+  while (val >= 100) {
+    uint64_t r = 2 * (val % 100);
+    val /= 100;
+
+    *index-- = DIGITS[r + 1];
+    *index-- = DIGITS[r];
+  }
+
+  // val is in [0..99]
+  // Decode last 2 or 1 digits
+  if (val < 10) {
+    *index-- = '0' + val;
+  } else {
+    val *= 2;
+    *index-- = DIGITS[val + 1];
+    *index-- = DIGITS[val];
+  }
+
+  return;
+}
+
+CAMLprim value unboxed_caml_int32_to_string(int32_t val) {
+  uint32_t x = val < 0 ? (uint32_t) (- val) : (uint32_t) val;
+  int count = u32_fast_digit_count(x);
+  value res = caml_alloc_string(count + (val < 0));
+  char* buf = (char*) String_val(res);
+
+  // caml_alloc_string(n) allocates n + 1 bytes
+  buf[count + (val < 0)] = 0;
```

Comment:
Nit: this is not necessary; the runtime representation of strings guarantees that they are always terminated with a zero byte (see https://dev.realworldocaml.org/runtime-memory-layout.html#scrollNav-6). In fact, the number of bytes allocated by `caml_alloc_string(n)` is actually the smallest number _greater_ than `n` that is a multiple of the word size.

(I haven't looked at the PR itself; this is just a detail that jumped at me.)

Reply:
I didn't know that, thanks!